File: ios/Sources/Ratakierros/Networking/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case badStatus(Int, String?)
    case decoding(Error)
    case transport(Error)
    case unauthorized

    var errorDescription: String? {
        switch self {
        case .badStatus(_, let body): return body ?? "Pyyntö epäonnistui."
        case .decoding: return "Vastauksen lukeminen epäonnistui."
        case .transport(let e): return e.localizedDescription
        case .unauthorized: return "Kirjautuminen vanhentunut."
        }
    }
}

actor APIClient {
    private let baseURL: URL
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(baseURL: URL? = nil, session: URLSession = .shared) {
        if let baseURL {
            self.baseURL = baseURL
        } else {
            let str = (Bundle.main.object(forInfoDictionaryKey: "API_BASE") as? String)
                ?? "https://ratakierros.fi"
            self.baseURL = URL(string: str)!
        }
        self.session = session
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    // MARK: - Public

    func tracks(lat: Double?, lon: Double?, query: String? = nil) async throws -> [Track] {
        var comps = URLComponents(url: baseURL.appendingPathComponent("/api/tracks"),
                                  resolvingAgainstBaseURL: false)!
        var items: [URLQueryItem] = []
        if let lat { items.append(URLQueryItem(name: "lat", value: String(lat))) }
        if let lon { items.append(URLQueryItem(name: "lon", value: String(lon))) }
        if let query, !query.isEmpty { items.append(URLQueryItem(name: "q", value: query)) }
        if !items.isEmpty { comps.queryItems = items }
        return try await get(comps.url!, token: nil)
    }

    func records(trackId: Int64, token: String?) async throws -> RecordsResponse {
        let url = baseURL.appendingPathComponent("/api/tracks/\(trackId)/records")
        return try await get(url, token: token)
    }

    func logRun(trackId: Int64, seconds: Double, token: String) async throws {
        let url = baseURL.appendingPathComponent("/api/runs")
        let body = LogRunRequest(trackId: trackId, timeSeconds: seconds)
        let _: LogRunResponse = try await post(url, body: body, token: token)
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        let url = baseURL.appendingPathComponent("/api/auth/login")
        return try await post(url, body: LoginRequest(email: email, password: password), token: nil)
    }

    func register(email: String, displayName: String, password: String) async throws -> AuthResponse {
        let url = baseURL.appendingPathComponent("/api/auth/register")
        let body = RegisterRequest(email: email, displayName: displayName, password: password)
        return try await post(url, body: body, token: nil)
    }

    // MARK: - Internal

    private func get<T: Decodable>(_ url: URL, token: String?) async throws -> T {
        var req = URLRequest(url: url)
        req.httpMethod = "GET"
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        return try await perform(req)
    }

    private func post<Body: Encodable, T: Decodable>(_ url: URL, body: Body, token: String?) async throws -> T {
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        req.httpBody = try encoder.encode(body)
        return try await perform(req)
    }

    private func perform<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, resp): (Data, URLResponse)
        do {
            (data, resp) = try await session.data(for: req)
        } catch {
            throw APIError.transport(error)
        }
        guard let http = resp as? HTTPURLResponse else {
            throw APIError.badStatus(0, nil)
        }
        if http.statusCode == 401 { throw APIError.unauthorized }
        guard (200..<300).contains(http.statusCode) else {
            let body = (try? decoder.decode(APIErrorPayload.self, from: data))?.error
                ?? String(data: data, encoding: .utf8)
            throw APIError.badStatus(http.statusCode, body)
        }
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decoding(error)
        }
    }
}

File: android/app/src/main/kotlin/fi/ratakierros/network/ApiClient.kt
package fi.ratakierros.network

import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import fi.ratakierros.auth.AuthRepository
import fi.ratakierros.model.AuthResponse
import fi.ratakierros.model.LogRunRequest
import fi.ratakierros.model.LogRunResponse
import fi.ratakierros.model.LoginRequest
import fi.ratakierros.model.RecordsResponse
import fi.ratakierros.model.RegisterRequest
import fi.ratakierros.model.Track
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

class ApiClient(baseUrl: String) {
    private val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
    }

    private val client = OkHttpClient.Builder().build()

    private val service: Service = Retrofit.Builder()
        .baseUrl(baseUrl.trimEnd('/') + "/")
        .client(client)
        .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
        .build()
        .create(Service::class.java)

    suspend fun tracks(lat: Double?, lon: Double?, query: String? = null): List<Track> =
        service.tracks(lat, lon, query?.takeIf { it.isNotBlank() })

    suspend fun records(trackId: Long, token: String?): RecordsResponse =
        service.records(trackId, token?.let { "Bearer $it" })

    suspend fun logRun(trackId: Long, seconds: Double, token: String): LogRunResponse =
        service.logRun(LogRunRequest(trackId, seconds), "Bearer $token")

    suspend fun login(email: String, password: String): AuthResponse =
        service.login(LoginRequest(email, password))

    suspend fun register(email: String, displayName: String, password: String): AuthResponse =
        service.register(RegisterRequest(email, displayName, password))

    interface Service {
        @GET("api/tracks")
        suspend fun tracks(
            @Query("lat") lat: Double?,
            @Query("lon") lon: Double?,
            @Query("q") q: String?,
        ): List<Track>

        @GET("api/tracks/{id}/records")
        suspend fun records(
            @Path("id") id: Long,
            @Header("Authorization") authorization: String?,
        ): RecordsResponse

        @POST("api/runs")
        suspend fun logRun(
            @Body body: LogRunRequest,
            @Header("Authorization") authorization: String,
        ): LogRunResponse

        @POST("api/auth/login")
        suspend fun login(@Body body: LoginRequest): AuthResponse

        @POST("api/auth/register")
        suspend fun register(@Body body: RegisterRequest): AuthResponse
    }
}
